### PR TITLE
Fixed of not exist ChoiceList  in Symfony 3+

### PR DESCRIPTION
**Pull request by eugenekurasov**
Closes [#691](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/issues/691
)

## Changelog

```markdown

### Fixed
Fixed ClassFilter for Symfony 3+. Remove deprecated for Symfony > 2.7

### CODE REVIEW

File: Filter/ClassFilter.php [hunk 1]
```diff
@@ -66,13 +66,24 @@ public function getFieldType()
...
     public function getFieldOptions()
     {
-        return $this->getOption('choices', array(
+        if (!class_exists('Symfony\Component\Form\Extension\Core\ChoiceList\ChoiceList')) {
+            $choiceOption = array(
+                'choices' => $this->getOption('sub_classes'),
+                'choices_as_values' => true, //NEXT_MAJOR: Remove this line when drop Symfony <3.0 support
+            );
+        } else {
+            //NEXT_MAJOR: Remove this condition when drop Symfony <2.7 support
```
Comment (greg0ire) [inline Filter/ClassFilter.php]: "dropping"

File: Filter/ClassFilter.php [hunk 2]
```diff
@@ -66,13 +66,24 @@ public function getFieldType()
      */
     public function getFieldOptions()
     {
-        return $this->getOption('choices', array(
+        if (!class_exists('Symfony\Component\Form\Extension\Core\ChoiceList\ChoiceList')) {
+            $choiceOption = array(
+                'choices' => $this->getOption('sub_classes'),
+                'choices_as_values' => true, //NEXT_MAJOR: Remove this line when drop Symfony <3.0 support
```
Comment (greg0ire) [inline Filter/ClassFilter.php]: "dropping"

File: Filter/ClassFilter.php [hunk 3]
```diff
@@ -66,13 +66,24 @@ public function getFieldType()
...
+            );
+        } else {
+            //NEXT_MAJOR: Remove this condition when drop Symfony <2.7 support
+            $choiceOption = array(
+                'choice_list' => new ChoiceList(
```
Comment (greg0ire) [inline Filter/ClassFilter.php]: Could you please remove `ChoiceList` altogether and do that kind of code instead: https://github.com/sonata-project/SonataAdminBundle/blob/025e8e1092ae80d305dcbeb817efb17e4ee7a4d2/Form/Type/Filter/DateTimeRangeType.php#L77

That should rid people still using sf 2.8 from some unwanted deprecation notices.

File: Filter/ClassFilter.php (hunk 1 above)
Comment (eugenekurasov) [inline Filter/ClassFilter.php]: We have discussion about this comment in [#691](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/issues/691#issuecomment-311638009). The goal of the comments is to make a note why this needed. You are want that I dropping this comment?

File: Filter/ClassFilter.php (hunk 2 above)
Comment (eugenekurasov) [inline Filter/ClassFilter.php]: We have discussion about this comment in [#691](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/issues/691#issuecomment-311638009). The goal of the comments is to make a note why this needed. You are want that I dropping this comment?

File: Filter/ClassFilter.php (hunk 3 above)
Comment (eugenekurasov) [inline Filter/ClassFilter.php]: Thank you for example.
I will update pull request tomorrow.

File: Filter/ClassFilter.php (hunk 2 above)
Comment (greg0ire) [inline Filter/ClassFilter.php]: No lol. I am suggesting you replace "drop" with "dropping"

**Comment by eugenekurasov** [comment]
@greg0ire Please review again

**Comment by eugenekurasov** [comment]
Sorry for a lot changing of CS. I need more sleep )

**Comment by jlamur** [comment]
You should not mention the issue in the commit title. I think it does not belong there, and it's polluting the issue everytime you modify your commit https://github.com/sonata-project/SonataDoctrineORMAdminBundle/issues/691#ref-commit-922d33d.

**Comment by greg0ire** [comment]
Please format your commit message according to our rules : 

- The commit message subject must be less than 50 characters and tell us *what* you did.
- The commit message body should tell us *why* you did it. It is optional but highly recommended.

Bad example :

```
Fixed bug #989 by removing call to defraculate()
```

Good example:

```
Remove call to defraculate()

Calling this function caused a bug because it interferes 
with calls to getSchmeckles().
Fixes #989
```

More details in CONTRIBUTING.md

File: Filter/ClassFilter.php [hunk 4]
```diff
@@ -13,10 +13,11 @@
 
 use Sonata\AdminBundle\Datagrid\ProxyQueryInterface;
 use Sonata\CoreBundle\Form\Type\EqualType;
-use Symfony\Component\Form\Extension\Core\ChoiceList\ChoiceList;
 
 class ClassFilter extends Filter
 {
+    const OPTION_SUB_CLASSES = 'sub_classes';
```
Comment (greg0ire) [inline Filter/ClassFilter.php]: Is there any benefit to making this publicly accessible (via `ClassFilter::OPTION_SUB_CLASSES`)?
Do we usually expose options with a constant?

Comment (eugenekurasov) [inline Filter/ClassFilter.php]: I used literal "sub_classes" twice. So I moved "sub_classes" to constant. It is more flexibility than use literal. If you want I do remove it.

Comment (jlamur) [inline Filter/ClassFilter.php]: No `const` in any other `Filter` class. It could be better to use an `OptionResolver` for options. Anyway, this does not belong in this PR.

Comment (eugenekurasov) [inline Filter/ClassFilter.php]: For example [sonar](https://www.sonarqube.org/) is throw warning if he detected use twice literal.

Comment (eugenekurasov) [inline Filter/ClassFilter.php]: @jlamur you are right, this code not belong in this PR. I will remove const.

Comment (greg0ire) [inline Filter/ClassFilter.php]: @eugenekurasov ok, but this pollutes the global scope. If you want to avoid repeating literals, please just use a variable (technically, a variable that never changes is... a constant). Note that with php 7.1 you could make the constant private, which would be a bit better, but would still make the constant available to the rest of the class.

Comment (greg0ire) [inline Filter/ClassFilter.php]: To be clear, what's not great about this is that the const will no longer be needed in next major when we simplify this, but removing it will be a BC-break. Removing a local variable isn't.

**Comment by eugenekurasov** [comment]
Changed commit message and removed const.
@greg0ire and @jlamur please review again.

**Comment by greg0ire** [comment]
Great job!

**Comment by jlamur** [comment]
Wouldn't it be cleaner to use the `choice_loader' option for SF 3+ instead of *tricky* functions checks on different classes / interfaces? Not a requirement but IMO easier to understand.

**Comment by greg0ire** [comment]
TIL about `choice_loader`, but the docs say: 

> The choice_loader can be used to only partially load the choices in cases where a fully-loaded list is not necessary. This is only needed in advanced cases and would replace the choices option.

I don't think our use case matches, because people hopefully don't have lots of subclasses.

**Comment by eugenekurasov** [comment]
"choice_loader" is not "choices". I think class filter should be same in different versions. PR is for fix ClassFilter

**Comment by greg0ire** [comment]
Also, bear in mind that on master, we are going to remove these tricks, and we should end up with just a `choices` option, which will be more straightforward. And the closer we are to the final state of the code, the better IMO.

**Comment by jlamur** [comment]
@greg0ire @eugenekurasov You're both right. PR is ok for me. :)

**Comment by eugenekurasov** [comment]
@jlamur if all ok, can you give approved

**Comment by greg0ire** [comment]
Thanks @eugenekurasov !

**Comment by OskarStark** [comment]
Released in [3.1.6](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/releases/tag/3.1.6)
